### PR TITLE
Format empty lines in stub files like black's preview style

### LINKED ISSUE #6723: Formatter: Empty lines in stub files
When formatting typeshed (`cargo run --bin ruff_dev -- format-dev --write typeshed`), there is still a big diff with black wrt to empty lines.

E.g. given
```python
class Feature:
    def getMandatoryRelease(self) -> None: ...
    compiler_flag: int

def getMandatoryRelease(self) -> None: ...
compiler_flag: int

if sys.platform != "win32":
    def can_change_color() -> bool: ...
    # Changed in Python 3.8.8 and 3.9.2
    if sys.version_info >= (3, 8):
        pass

if sys.platform == "win32":
    class Dummy:
        __new__: None
        __init__: None

    # Actual typename SummaryInformation, not exposed by the implementation
    class _SummaryInformation:
        def GetProperty(self, field: int) -> int | bytes | None: ...
```
black will format this as
```python
class Feature:
    def getMandatoryRelease(self) -> None: ...
    compiler_flag: int

def getMandatoryRelease(self) -> None: ...

compiler_flag: int

if sys.platform != "win32":
    def can_change_color() -> bool: ...
    # Changed in Python 3.8.8 and 3.9.2
    if sys.version_info >= (3, 8):
        pass

if sys.platform == "win32":
    class Dummy:
        __new__: None
        __init__: None

    # Actual typename SummaryInformation, not exposed by the implementation
    class _SummaryInformation:
        def GetProperty(self, field: int) -> int | bytes | None: ...
```
while we format this as
```python
class Feature:
    def getMandatoryRelease(self) -> None: ...

    compiler_flag: int

def getMandatoryRelease(self) -> None: ...

compiler_flag: int

if sys.platform != "win32":

    def can_change_color() -> bool: ...

    # Changed in Python 3.8.8 and 3.9.2
    if sys.version_info >= (3, 8):
        pass

if sys.platform == "win32":

    class Dummy:
        __new__: None
        __init__: None

    # Actual typename SummaryInformation, not exposed by the implementation
    class _SummaryInformation:
        def GetProperty(self, field: int) -> int | bytes | None: ...
```

The task is to go through the typeshed diff and figure out each empty line rule we're still missing. If you implement this you don't need to (and probably shouldn't) fix everything at once but you can implement any part of it (as long as it doesn't decrease the similarity index).

### PR DESCRIPTION
**Pull request by konstin**
## Summary

Fix all but one empty line differences with the black preview style in typeshed. The remaining differences are breaking with type comments and trailing commas in function definitions.

I compared the empty line differences with the preview mode of black since stable has some oddities that would have been hard to replicate (https://github.com/psf/black/issues/3861). Additionally, it assumes the style proposed in https://github.com/psf/black/issues/3862.

An edge case that also surfaced with typeshed are newline before trailing module comments.

**main**

| project      | similarity index  | total files       | changed files     |
|--------------|------------------:|------------------:|------------------:|
| cpython      |           0.76083 |              1789 |              1632 |
| django       |           0.99966 |              2760 |                58 |
| transformers |           0.99930 |              2587 |               447 |
| twine        |           1.00000 |                33 |                 0 |
| **typeshed**     |           0.99978 |              3496 |              **2173** |
| warehouse    |           0.99825 |               648 |                22 |
| zulip        |           0.99950 |              1437 |                27 |

**PR**
| project      | similarity index  | total files       | changed files     |
|--------------|------------------:|------------------:|------------------:|
| cpython      |           0.76083 |              1789 |              1632 |
| django       |           0.99966 |              2760 |                58 |
| transformers |           0.99930 |              2587 |               447 |
| twine        |           1.00000 |                33 |                 0 |
| **typeshed**     |           0.99983 |              3496 |                **18** |
| warehouse    |           0.99825 |               648 |                22 |
| zulip        |           0.99950 |              1437 |                27 |

Closes #6723

## Test Plan

The main driver was the typeshed diff. I added new test cases for all kinds of possible empty line combinations in stub files, test cases for newlines before trailing module comments.

### CODE REVIEW

File: crates/ruff_python_formatter/src/statement/stmt_class_def.rs [hunk 1]
```diff
@@ -128,15 +128,13 @@ impl FormatNodeRule<StmtClassDef> for FormatStmtClassDef {
...
-            } else {
-                1
-            },
-        )
-        .fmt(f)
+        let empty_lines = match (f.options().source_type(), f.context().node_level()) {
+            (PySourceType::Stub, NodeLevel::TopLevel) => 1,
+            (PySourceType::Stub, _) => 0,
+            (_, NodeLevel::TopLevel) => 2,
+            (_, _) => 1,
```
Comment (konstin) [inline crates/ruff_python_formatter/src/statement/stmt_class_def.rs:135]: it's a bit counterintuitive that the exceptions are the named cases

Comment (MichaReiser) [reply crates/ruff_python_formatter/src/statement/stmt_class_def.rs:135]: Should `empty_lines` be inlined into `empty_lines_before_trailing_comments`. It seems the function is only used in `stmt_function_def` and `stmt_class_def` and both use the same implementation.

File: crates/ruff_python_formatter/src/statement/suite.rs [hunk 2]
```diff
@@ -345,6 +327,89 @@ impl FormatRule<Suite, PyFormatContext<'_>> for FormatSuite {
...
+                empty_line().fmt(f)
+            } else {
+                hard_line_break().fmt(f)
+            }
+        }
+    }
+}
+
+/// Only a function to compute it lazily
+fn stub_suite_can_omit_empty_line(preceding: &Stmt, following: &Stmt, f: &mut PyFormatter) -> bool {
```
Comment (MichaReiser) [inline crates/ruff_python_formatter/src/statement/suite.rs:366]: Can we make the formatter readonly. The function doesn't seem to call into `fmt`
```suggestion
fn stub_suite_can_omit_empty_line(preceding: &Stmt, following: &Stmt, f: &PyFormatter) -> bool {
```

**Review by MichaReiser — APPROVED** [review]

File: crates/ruff_python_formatter/src/statement/suite.rs (hunk 2 above)
Comment (konstin) [reply crates/ruff_python_formatter/src/statement/suite.rs:366]: huh thanks good catch, strange that clippy didn't pick that up

**Pull request merged**